pkg/injector/pod_patch.go

```diff
@@ -759,21 +793,35 @@ func getSidecarContainer(annotations map[string]string, id, daprSidecarImage, im
 		},
 		ReadinessProbe: &corev1.Probe{
 			ProbeHandler:        httpHandler,
-			InitialDelaySeconds: getInt32AnnotationOrDefault(annotations, daprReadinessProbeDelayKey, defaultHealthzProbeDelaySeconds),
-			TimeoutSeconds:      getInt32AnnotationOrDefault(annotations, daprReadinessProbeTimeoutKey, defaultHealthzProbeTimeoutSeconds),
-			PeriodSeconds:       getInt32AnnotationOrDefault(annotations, daprReadinessProbePeriodKey, defaultHealthzProbePeriodSeconds),
-			FailureThreshold:    getInt32AnnotationOrDefault(annotations, daprReadinessProbeThresholdKey, defaultHealthzProbeThreshold),
+			InitialDelaySeconds: getInt32AnnotationOrDefault(cfg.annotations, daprReadinessProbeDelayKey, defaultHealthzProbeDelaySeconds),
+			TimeoutSeconds:      getInt32AnnotationOrDefault(cfg.annotations, daprReadinessProbeTimeoutKey, defaultHealthzProbeTimeoutSeconds),
+			PeriodSeconds:       getInt32AnnotationOrDefault(cfg.annotations, daprReadinessProbePeriodKey, defaultHealthzProbePeriodSeconds),
+			FailureThreshold:    getInt32AnnotationOrDefault(cfg.annotations, daprReadinessProbeThresholdKey, defaultHealthzProbeThreshold),
 		},
 		LivenessProbe: &corev1.Probe{
 			ProbeHandler:        httpHandler,
-			InitialDelaySeconds: getInt32AnnotationOrDefault(annotations, daprLivenessProbeDelayKey, defaultHealthzProbeDelaySeconds),
-			TimeoutSeconds:      getInt32AnnotationOrDefault(annotations, daprLivenessProbeTimeoutKey, defaultHealthzProbeTimeoutSeconds),
-			PeriodSeconds:       getInt32AnnotationOrDefault(annotations, daprLivenessProbePeriodKey, defaultHealthzProbePeriodSeconds),
-			FailureThreshold:    getInt32AnnotationOrDefault(annotations, daprLivenessProbeThresholdKey, defaultHealthzProbeThreshold),
+			InitialDelaySeconds: getInt32AnnotationOrDefault(cfg.annotations, daprLivenessProbeDelayKey, defaultHealthzProbeDelaySeconds),
+			TimeoutSeconds:      getInt32AnnotationOrDefault(cfg.annotations, daprLivenessProbeTimeoutKey, defaultHealthzProbeTimeoutSeconds),
+			PeriodSeconds:       getInt32AnnotationOrDefault(cfg.annotations, daprLivenessProbePeriodKey, defaultHealthzProbePeriodSeconds),
+			FailureThreshold:    getInt32AnnotationOrDefault(cfg.annotations, daprLivenessProbeThresholdKey, defaultHealthzProbeThreshold),
 		},
 	}
 
-	c.Env = append(c.Env, utils.ParseEnvString(annotations[daprEnvKey])...)
+	// This is a special case for virtual kubelets on Azure Container Instances (ACI).
+	// ACI does not support separate Command and Args,
+	// Ref. https://github.com/virtual-kubelet/azure-aci/issues/10.
+	// For other deployments, the Command is passed as a part of Args.
+	// This allows the Docker images to specify an ENTRYPOINT
+	// which is otherwise overridden by Command.
+	if isACIVirtualKubelet(cfg.tolerations) {
```

Comment:
I like that we found a solution for these cases. Thanks, @shubham1172! I think we can avoid having "if cloud provider" logic in the code here. We can change it to receive the list of tolerations from an environment variable and have the list to come from the Helm chart. The list would still be hardcoded, but in the Helm chart instead. This way, other cloud providers living the same situation can add to the list of tolerations in the Helm chart. I don't think that the list should be configurable by the end user since it gives a better experience out of the box without requiring user to have a problem and then discover the documentation about which toleration to add.

Reply:
Thanks @artursouza, updated! The default list contains just one toleration for ACI, and can be updated via Helm charts to include more, they are matched with pod tolerations using logical AND.